crates/sargon/src/profile/mfa/security_structures/automatic_shield_builder/automatic_shield_builder.rs

```diff
@@ -0,0 +1,916 @@
+    ) -> Result<AutoBuildOutcomeForTesting>
+/* Feel free to replace `AutoBuildOutcomeForTesting` return type if you need anything else, I had Unit, so might as well make testing easier by returning this type. */
+    {
+        if let Some(invalid_reason) = self.validate_primary_role() {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: format!(
+                    "Primary role is not valid: {:?}",
+                    invalid_reason
+                ),
+            });
+        }
+
+        if !self.get_primary_override_factors().is_empty() {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: "Primary override factors not allowed when preselecting factors for Recovery and Confirmation".to_string(),
+            });
+        }
+
+        let primary_factors = self
+            .get_primary_threshold_factors()
+            .into_iter()
+            .collect::<IndexSet<_>>();
+
+        if primary_factors
+            .intersection(
+                &all_factors_in_profile
+                    .iter()
+                    .map(|f| f.id())
+                    .collect::<IndexSet<_>>(),
+            )
+            .cloned()
+            .collect::<IndexSet<FactorSourceID>>()
+            != primary_factors
+        {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: "Primary factors not in profile".to_string(),
+            });
+        }
+
+        if !Self::prerequisites_status(
+            &all_factors_in_profile.iter().map(|f| f.id()).collect(),
+        )
+        .is_sufficient()
+        {
+            return Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: "Prerequisites not met".to_string(),
+            });
+        }
+
+        let mut auto_builder = AutomaticShieldBuilder::new(
+            all_factors_in_profile,
+            primary_factors,
+        );
+
+        let proto_matrix = auto_builder.assign()?;
+
+        assert_eq!(
+            proto_matrix.primary.clone().into_iter().collect_vec(),
+            self.get_primary_threshold_factors(),
+            "Auto assignment should not have changed the primary factors"
+        );
+        self.set_state(proto_matrix);
+
+        if let Some(invalid_reason) = self.validate() {
+            Err(CommonError::AutomaticShieldBuildingFailure {
+                underlying: invalid_reason.to_string(),
+            })
+        } else {
+            Ok(auto_builder.stats_for_testing)
+        }
+    }
+
+    /// Updates the Primary, Recovery and Confirmation roles with the factors of the given `ProtoMatrix`.
+    fn set_state(&self, proto_matrix: ProtoMatrix) {
+        self.reset_factors_in_roles();
+        self.set_threshold(proto_matrix.primary.len() as u8);
+        proto_matrix.primary.into_iter().for_each(|f| {
+            self.add_factor_source_to_primary_threshold(f);
+        });
```

Comment:
Can you elaborate on this part?

Reply:
`ProtoMatrix` is just a named tuple of the three lists of factors, for the three roles.

The `auto_assign_factors_to_recovery_and_confirmation_based_on_primary` either is fully successful, or fails - without any mutation. So the assignment of the factors to role need to happen NOT to `self` (`SecurityShieldBuilder`) so they happen to `ProtoMatrix`.

Once assignment is successful `SecurityShieldBuilder` instance calls `set_state` to set the factors.